File: app/src/main/java/com/jrvermeer/psalter/Infrastructure/Logger.java
package com.jrvermeer.psalter.Infrastructure;

import android.content.Context;
import android.util.Log;

import com.flurry.android.FlurryAgent;
import com.jrvermeer.psalter.Core.Models.LogEvent;
import com.jrvermeer.psalter.Core.Models.SearchMode;
import com.jrvermeer.psalter.R;
import com.jrvermeer.psalter.UI.PsalterApplication;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Jonathan on 7/17/2018.
 */

public class Logger {
    private Logger() { }

    public static void d(String message){
        Log.d("PsalterLog", message);
    }

    public static void e(String message, Throwable ex){
        Log.e("PsalterLog", message, ex);
    }

    static {
        Context c =  PsalterApplication.getContext();
        new FlurryAgent.Builder()
                .withLogEnabled(true)
                .build(c, c.getResources().getString(R.string.secret_flurry));
    }
    public static void changeScore(boolean scoreVisible){
        Map<String, String> params = new HashMap<>();
        params.put("ScoreVisible", String.valueOf(scoreVisible));
        FlurryAgent.logEvent(LogEvent.ChangeScore.name(), params);
    }
    public static void changeTheme(boolean nightMode){
        Map<String, String> params = new HashMap<>();
        params.put("NightMode", String.valueOf(nightMode));
        FlurryAgent.logEvent(LogEvent.ChangeTheme.name(), params);
    }
    public static void playbackStarted(String numberTitle, boolean shuffling){
        Map<String, String> params = new HashMap<>();
        params.put("Number", numberTitle);
        FlurryAgent.logEvent(shuffling ? LogEvent.ShufflePsalter.name() : LogEvent.PlayPsalter.name(), params);
    }

    public static void searchEvent(SearchMode mode, String query, String psalterChosen){
        Map<String, String> params = new HashMap<>();

        LogEvent event;
        if (mode == SearchMode.Psalter){
            event  = LogEvent.SearchPsalter;
            params.put("Psalter", query);
        }
        else {
            if(mode == SearchMode.Psalm){
                event = LogEvent.SearchPsalm;
                params.put("Psalm", query);
            }
            else { // SearchMode.Lyrics
                event = LogEvent.SearchLyrics;
                params.put("Query", query);
            }
            params.put("PsalterChosen", psalterChosen);
        }
        FlurryAgent.logEvent(event.name(), params);
    }

    public static void event(LogEvent event){
        FlurryAgent.logEvent(event.name());
    }
    public static void error(Throwable ex){
        FlurryAgent.onError("Error", "", ex);
    }
    public static void error(String message){
        FlurryAgent.onError("Error", message, "");
    }
}

File: app/src/main/java/com/jrvermeer/psalter/infrastructure/Logger.kt
package com.jrvermeer.psalter.infrastructure

import android.content.Context
import android.util.Log

import com.flurry.android.FlurryAgent
import com.jrvermeer.psalter.models.LogEvent
import com.jrvermeer.psalter.models.SearchMode
import com.jrvermeer.psalter.R
import com.jrvermeer.psalter.models.Psalter

/**
 * Created by Jonathan on 7/17/2018.
 */

object Logger {
    private const val TAG = "PsalterLog"

    fun init(context: Context) {
        FlurryAgent.Builder()
                .withLogEnabled(true)
                .build(context, context.resources.getString(R.string.secret_flurry))
    }

    fun d(message: String) {
        Log.d(TAG, message)
    }

    fun e(message: String, ex: Throwable?) {
        Log.e(TAG, message, ex)
        FlurryAgent.onError("Error", message, ex)
    }

    fun changeScore(scoreVisible: Boolean) {
        val params = mapOf("ScoreVisible" to scoreVisible.toString())
        event(LogEvent.ChangeScore, params)
    }

    fun changeTheme(nightMode: Boolean) {
        val params = mapOf("NightMode" to nightMode.toString())
        event(LogEvent.ChangeTheme, params)
    }

    fun playbackStarted(numberTitle: String, shuffling: Boolean) {
        val params = mapOf("Number" to numberTitle, "Shuffling" to shuffling.toString())
        event(LogEvent.PlayPsalter, params)
    }

    fun skipToNext(skippedPsalter: Psalter?){
        val params = mapOf("SkippedPsalter" to (skippedPsalter?.title ?: "null"))
        event(LogEvent.SkipToNext, params)
    }

    fun searchPsalter(number: Int){
        event(LogEvent.SearchPsalter, mapOf("Number" to number.toString()))
    }
    fun searchPsalm(psalm: Int, psalterChosen: Int? = null){
        val params = mutableMapOf("Psalm" to psalm.toString())
        if(psalterChosen != null) params["PsalterChosen"] = psalterChosen.toString()
        event(LogEvent.SearchPsalm, params)
    }
    fun searchLyrics(query: String, psalterChosen: Int? = null){
        val params = mutableMapOf("Query" to query)
        if(psalterChosen != null) params["PsalterChosen"] = psalterChosen.toString()
        event(LogEvent.SearchLyrics, params)
    }
    fun searchEvent(searchMode: SearchMode, query: String, psalterChosen: Int? = null){
        when (searchMode){
            SearchMode.Lyrics -> searchLyrics(query, psalterChosen)
            SearchMode.Psalter -> searchPsalter(query.toInt())
            SearchMode.Psalm -> searchPsalm(query.toInt(), psalterChosen)
            SearchMode.Favorites -> event(LogEvent.SearchFavorites, mutableMapOf("PsalterChosen" to psalterChosen.toString()))
        }
    }
    fun ratePromptAttempt(timesShown: Int){
        val params = mutableMapOf("TimesShown" to timesShown.toString())
        event(LogEvent.RateFlow, params)
    }

    fun event(event: LogEvent, params: Map<String, String>? = null) {
        FlurryAgent.logEvent(event.name, params)
    }
}
